server/src/engineModel/EngineModel.ts

```diff
@@ -1,97 +1,186 @@
+    }
 
-/**
- * Add a new class id to the completion list as a new completion item.
- * @param classyid The id of class to add as a completion item.
- * @param completionItems The completion items list.
- */
-function addYclassCompletionItem(
-  uri: string,
-  classyid: string,
-  completionProvider: YmlCompletionItemsProvider,
-): void {
-  const itemId = `id_${classyid}`;
-  if (
-    !completionProvider.getItem(itemId)
-  ) {
-    const completionItem = {
-      data: itemId,
-      detail: `Yseop Engine type: “${classyid}”.`,
-      // ,documentation: "Its documentation can come from predefinedObjects.xml"
-      kind: CompletionItemKind.Class,
-      label: classyid,
-    };
-    completionProvider.addCompletionItem({uri, completion: completionItem});
-  }
+    private buildAttribute(attributeXmlElement: any, sourceType: string): YmlAttribute {
+        const attribute = new YmlAttribute(attributeXmlElement.$.ident, this.uri);
+        attribute.detail = `[${sourceType}].${attribute.label}`;
+        attributeXmlElement.return.forEach((returnType) => {
+            if (returnType.domains) {
+                attribute.domains = returnType.domains;
+            }
+            if (returnType.domainsLevel2) {
+                attribute.domainsLevel2 = returnType.domainsLevel2;
+            }
+        });
+        return attribute;
+    }
+
+    private parsePredefinedObjects(uri: string): void {
+        connection.console.log(`Parsing definition file: ${uri}`);
+        fs.readFile(uri, (err, data) => {
+            parser.parseString(data, (parseErr, predefinedObjects) => {
+                if (err != null) {
+                    connection.console.error(`Something went wrong during YE model import:\n ${parseErr}`);
+                } else if (predefinedObjects == null) {
+                    connection.console.error('Something went wrong during YE model import. Your file seems empty.');
+                } else {
+                    try {
+                        this.importClasses(predefinedObjects);
+                        this.importFunctions(predefinedObjects);
+                        this.importTags(predefinedObjects);
+                    } catch (importErr) {
+                        connection.console.error(`Something went wrong during YE model import:\n ${importErr}`);
+                    }
+                }
+                connection.console.log(
+                    `Done with classes size=${this.classes.length}\nfunctions size=${this.functions.length}`,
+                );
+            });
+        });
+    }
+
+    /**
+     * Import all the YML classes.
+     *
+     * @param predefinedObjects The data structure resulting of parsing `predefinedObjects.xml`.
+     */
+    private importClasses(predefinedObjects: any): void {
+        connection.console.log('Importing classes from Yseop Engine model.');
+        const classesByPackage = predefinedObjects['data-and-features'].classes[0].package;
```

Comment:
I don't know what the `[0]` is targetting and I'd like to have a comment explaining that.